File: src/services/withdrawals.ts
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8000';

// Types
export interface WithdrawalRequest {
  amount: number;
  account_type: 'CBE' | 'TeleBirr';
  account_number: string;
  account_holder_name: string;
  phone_number?: string;
}

export interface WithdrawalResponse {
  id: string;
  userId: string;
  amount: number;
  account_type: string;
  account_number: string;
  account_holder_name: string;
  phone_number?: string;
  status: 'pending' | 'approved' | 'rejected';
  admin_notes?: string;
  rejection_reason?: string;
  created_at: string;
  processed_at?: string;
  processed_by?: string;
  // Add user info for admin view
  userName?: string;
  userEmail?: string;
}

export interface WithdrawalStats {
  totalEarnings: number;
  totalWithdrawn: number;
  availableBalance: number;
  pendingWithdrawals: number;
}

class WithdrawalsService {
  private getAuthHeaders() {
    const token = localStorage.getItem('access_token');
    return {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json'
    };
  }

  private getAdminAuthHeaders() {
    const token = localStorage.getItem('adminToken');
    return {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json'
    };
  }

  async createWithdrawalRequest(withdrawalData: WithdrawalRequest): Promise<WithdrawalResponse> {
    try {
      const response = await axios.post(
        `${API_BASE_URL}/withdrawals/`,
        withdrawalData,
        { headers: this.getAuthHeaders() }
      );
      return response.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to create withdrawal request');
    }
  }

  async getMyWithdrawals(): Promise<WithdrawalResponse[]> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/withdrawals/my`,
        { headers: this.getAuthHeaders() }
      );
      return response.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch withdrawal requests');
    }
  }

  async getWithdrawalStats(): Promise<WithdrawalStats> {
    try {
      // Get user profile to extract earnings info
      const response = await axios.get(
        `${API_BASE_URL}/dashboard/stats`,
        { headers: this.getAuthHeaders() }
      );
      
      const stats = response.data;
      const withdrawals = await this.getMyWithdrawals();
      
      const totalWithdrawn = withdrawals
        .filter(w => w.status === 'approved')
        .reduce((sum, w) => sum + w.amount, 0);
      
      const pendingWithdrawals = withdrawals
        .filter(w => w.status === 'pending')
        .reduce((sum, w) => sum + w.amount, 0);

      return {
        totalEarnings: stats.totalEarnings || 0,
        totalWithdrawn,
        availableBalance: (stats.totalEarnings || 0) - totalWithdrawn,
        pendingWithdrawals
      };
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch withdrawal stats');
    }
  }

  async getAllWithdrawals(): Promise<WithdrawalResponse[]> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/withdrawals/`,
        { headers: this.getAdminAuthHeaders() }
      );
      return response.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to fetch all withdrawal requests');
    }
  }

  async approveWithdrawal(withdrawalId: string, admin_notes?: string): Promise<void> {
    try {
      const url = admin_notes 
        ? `${API_BASE_URL}/withdrawals/${withdrawalId}/approve?admin_notes=${encodeURIComponent(admin_notes)}`
        : `${API_BASE_URL}/withdrawals/${withdrawalId}/approve`;
      
      await axios.post(
        url,
        {},
        { headers: this.getAdminAuthHeaders() }
      );
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to approve withdrawal request');
    }
  }

  async rejectWithdrawal(withdrawalId: string, rejection_reason: string): Promise<void> {
    try {
      await axios.post(
        `${API_BASE_URL}/withdrawals/${withdrawalId}/reject?rejection_reason=${encodeURIComponent(rejection_reason)}`,
        {},
        { headers: this.getAdminAuthHeaders() }
      );
    } catch (error: any) {
      throw new Error(error.response?.data?.message || 'Failed to reject withdrawal request');
    }
  }
}

export const withdrawalsService = new WithdrawalsService();

File: backend/routes/withdrawals.py
from fastapi import APIRouter, HTTPException, status, Depends
from typing import List
from models import WithdrawalRequest, WithdrawalResponse
from auth import get_current_user
from routes.admin import get_current_admin
from database.operations import db_ops

router = APIRouter(prefix="/withdrawals", tags=["Withdrawals"])

@router.post("/", response_model=WithdrawalResponse)
async def create_withdrawal_request(
    withdrawal_data: WithdrawalRequest,
    current_user: dict = Depends(get_current_user)
):
    """Create a withdrawal request"""
    try:
        # Validate minimum amount
        if withdrawal_data.amount < 300:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Minimum withdrawal amount is 300 ETB"
            )
        
        # Get user's current earnings
        user_profile = await db_ops.get_user_by_id(current_user["id"])
        if not user_profile:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User profile not found"
            )
        
        current_earnings = float(user_profile.get("total_earnings", 0))
        
        # Check if user has enough earnings
        if current_earnings < withdrawal_data.amount:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Insufficient earnings. Available: {current_earnings} ETB, Requested: {withdrawal_data.amount} ETB"
            )
        
        # Create withdrawal request
        withdrawal = await db_ops.create_withdrawal_request(
            current_user["id"], 
            withdrawal_data.dict()
        )
        
        if not withdrawal:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to create withdrawal request"
            )
        
        # Map database fields to API response fields
        mapped_withdrawal = {
            "id": withdrawal["id"],
            "userId": withdrawal["user_id"],
            "amount": withdrawal["amount"],
            "accountType": withdrawal["account_type"],
            "accountNumber": withdrawal["account_number"],
            "accountHolderName": withdrawal["account_holder_name"],
            "phoneNumber": withdrawal.get("phone_number"),
            "status": withdrawal["status"],
            "adminNotes": withdrawal.get("admin_notes"),
            "rejectionReason": withdrawal.get("rejection_reason"),
            "createdAt": withdrawal["created_at"],
            "processedAt": withdrawal.get("processed_at"),
            "processedBy": withdrawal.get("processed_by")
        }
        
        return WithdrawalResponse(**mapped_withdrawal)
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create withdrawal request: {str(e)}"
        )

@router.get("/my", response_model=List[WithdrawalResponse])
async def get_my_withdrawal_requests(
    current_user: dict = Depends(get_current_user)
):
    """Get user's withdrawal requests"""
    try:
        withdrawals = await db_ops.get_user_withdrawal_requests(current_user["id"])
        # Map database fields to API response fields
        mapped_withdrawals = []
        for withdrawal in withdrawals:
            mapped_withdrawal = {
                "id": withdrawal["id"],
                "userId": withdrawal["user_id"],
                "amount": withdrawal["amount"],
                "accountType": withdrawal["account_type"],
                "accountNumber": withdrawal["account_number"],
                "accountHolderName": withdrawal["account_holder_name"],
                "phoneNumber": withdrawal.get("phone_number"),
                "status": withdrawal["status"],
                "adminNotes": withdrawal.get("admin_notes"),
                "rejectionReason": withdrawal.get("rejection_reason"),
                "createdAt": withdrawal["created_at"],
                "processedAt": withdrawal.get("processed_at"),
                "processedBy": withdrawal.get("processed_by")
            }
            mapped_withdrawals.append(WithdrawalResponse(**mapped_withdrawal))
        return mapped_withdrawals
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch withdrawal requests: {str(e)}"
        )

@router.get("/", response_model=List[WithdrawalResponse])
async def get_all_withdrawal_requests(
    status: str = None,
    current_admin: dict = Depends(get_current_admin)
):
    """Get all withdrawal requests (admin only)"""
    try:
        withdrawals = await db_ops.get_all_withdrawal_requests(status)
        # Map database fields to API response fields
        mapped_withdrawals = []
        for withdrawal in withdrawals:
            mapped_withdrawal = {
                "id": withdrawal["id"],
                "userId": withdrawal["user_id"],
                "amount": withdrawal["amount"],
                "accountType": withdrawal["account_type"],
                "accountNumber": withdrawal["account_number"],
                "accountHolderName": withdrawal["account_holder_name"],
                "phoneNumber": withdrawal.get("phone_number"),
                "status": withdrawal["status"],
                "adminNotes": withdrawal.get("admin_notes"),
                "rejectionReason": withdrawal.get("rejection_reason"),
                "createdAt": withdrawal["created_at"],
                "processedAt": withdrawal.get("processed_at"),
                "processedBy": withdrawal.get("processed_by")
            }
            mapped_withdrawals.append(WithdrawalResponse(**mapped_withdrawal))
        return mapped_withdrawals
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch withdrawal requests: {str(e)}"
        )

@router.post("/{withdrawal_id}/approve")
async def approve_withdrawal_request(
    withdrawal_id: str,
    admin_notes: str = None,
    current_admin: dict = Depends(get_current_admin)
):
    """Approve a withdrawal request (admin only)"""
    try:
        success = await db_ops.approve_withdrawal_request(
            withdrawal_id, 
            current_admin["id"], 
            admin_notes
        )
        
        if not success:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Failed to approve withdrawal request"
            )
        
        return {"message": "Withdrawal request approved successfully"}
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to approve withdrawal request: {str(e)}"
        )

@router.post("/{withdrawal_id}/reject")
async def reject_withdrawal_request(
    withdrawal_id: str,
    rejection_reason: str,
    current_admin: dict = Depends(get_current_admin)
):
    """Reject a withdrawal request (admin only)"""
    try:
        success = await db_ops.reject_withdrawal_request(
            withdrawal_id, 
            current_admin["id"], 
            rejection_reason
        )
        
        if not success:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Failed to reject withdrawal request"
            )
        
        return {"message": "Withdrawal request rejected successfully"}
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to reject withdrawal request: {str(e)}"
        )
